# ICS 434 Homework Assignment 1

# <font color="red">Due: 11:59 PM on Wednesday, February 12</font>

### Submission Instructions

1. Name your notebook file with the following naming convention using your last and first name as presented in Laulima.
    - [LastName][FirstName]_[AssignmentNumber].ipynb
    - For example, Bruce Wayne ==> WayneBruce_1.ipynb (The number at the end is the assignment number.)
2. Only use .ipynb file extension. Other extensions (file formats) like .rtf, .zip, .docs, .pdf are not accepted.
4. Never compress your files in a zip file. Data files are available to the instructor and the TA, so no need to upload them to Laulima. Make sure you use the same filenames of data files as given in the homework.
5. Save data files in **"data" folder under your working directory**. Use **relative path** when you read in data in your code.
6. Do not create any subfolders in your Drop Box.
7. **Do NOT modify or delete the provided code.**
8. Clean your code before submission.
    - If needed, provide clear documentation describing the purpose and how to use every class or function in your code.
    - Your submission **should show only the required outputs**.
9. Run your code before submission to **show all outputs in the submitted file**.
10. Write your full name in the cell below.
***


## Your Name: Lionel Derrick Roxas

***

### Brief Overview

- The goal of this assignment is to practice some important data wrangling functionality commonly required in real-world projects.

- Here we will use two datasets:
  - IRS Statistics of Income (SOI) dataset
  - The Medicaid Data per State  


- The final product of this assignment is a table with medication cost per Medicaid enrollee per state. The Medicaid dataset will allow us to answer questions such as:
  - Which medications account for the bulk of a state's spending   
  - Which drugs are prescribed much more in one state compared to the other states.
etc.



### Instructions on answering the questions

- Some of the questions below require that you only use methods or properties of a `DataFrame` or a `Series`. Therefore, any solution that uses a function that is not a method or property of `DataFrame` or `Series` will not be accepted, even when the solution yields an appropriate answer to the question.

- For instance, if you are asked to find the number of entries in the DataFrame `tax_data` using only the DataFrame's methods or properties, then `len(tax_data)` is not an acceptable solution since `len()` is not a `tax_data` method. The statements below are both correct answers:

> - `info()` is a method of `tax_data`.
  
```python
        tax_data.info()
```

&emsp;&emsp; or

> - `shape` is a property of `tax_data`.

```python
        tax_data.shape
```

- Similarly, if you are asked to count the number of unique entries in the `STATEFIPS` column of the `tax_data` DataFrame, then solutions using `set()` or `len()` are not acceptable for the following reasons:

> - The solution below uses `set()` and `len()`, which are not `tax_data` methods.

```python
        len(set(tax_data["STATEFIPS"]))
```

> - The solution below uses `unique()`, which is a  `tax_data` method, but then counts the number of unique entries using `len()` which is not `tax_data` method.

```python
        len(tax_data["STATEFIPS"].unique())
```


- The statement below is an acceptable solution since it uses `nunique()` which is a method of the `Series` generated by indexing on a column of `tax_data` (tax_data["STATEFIPS"]).

```python
    tax_data['STATEFIPS'].nunique()
```


- Chaining methods and properties is encouraged if it does not cause ambiguity.

- For instance, to identify whether a value is part of the index, write the following:

```python
    tax_data.index.contains(99999)
```

&emsp;&emsp; rather than:

```python
    9999 in tax_data.index
```

- You can only import `pandas` and `numpy`.

- If you are not explicitely asked to only use methods or properties of a `DataFrame` or a `Series`, then any solution that does not rely on external products will be accepted.

- __IMPORTANT__: Provide the exact statement(s) used to answer each question.

- Unless otherwise specified, each cell can contain multiple lines of code.

* Finally, note that not all the functions necessary for answering the questions below were covered in class. As such, I suggest you use `.SHIFT+TAB` or `.TAB` on objects liberally to see which methods and properties are available on objects. If you are unsure what a method does, use `SHIFT+TAB` to invoke the `docstring`, or documentation for that function. This is not only a good way to see which functionality can be used to answer the questions below but also a great way to familiarize yourself with the plethora of functionality available through the `pandas` package.

---

# 1. Working with the Tax Data

## 1.1. Loading and exploring the data

In [ ]:
import pandas as pd
import numpy as np


* Load the IRS Statistics of Income (SOI) dataset (`tax_data.csv`) into a DataFrame called `tax_data`. The file is located in the `data` directory of the assignment folder.

* This dataset was preprocessed but the original one was obtained at the following URL:

        https://www.irs.gov/pub/irs-soi/15zpallagi.csv


In [5]:
## WRITE YOUR CODE HERE

tax_data = pd.read_csv('data/tax_data.csv')

* Use a `tax_data` method or property to display the first eight (8) rows of the `DataFrame`.

In [45]:
## WRITE YOUR CODE HERE

tax_data.head(8)

STATEFIPS STATE  ZIPCODE  AGI_STUB        N1     MARS1     MARS2     MARS4  \
0        1.0    AL      0.0       1.0  836320.0  481570.0  109790.0  233260.0   
1        1.0    AL      0.0       2.0  494830.0  206630.0  146250.0  129390.0   
2        1.0    AL      0.0       3.0  261250.0   80720.0  139280.0   36130.0   
3        1.0    AL      0.0       4.0  166690.0   28510.0  124650.0   10630.0   
4        1.0    AL      0.0       5.0  212660.0   19520.0  184320.0    4830.0   
5        1.0    AL      0.0       6.0   55360.0    2950.0   49260.0     350.0   
6        1.0    AL  35004.0       1.0    1490.0     970.0     230.0     280.0   
7        1.0    AL  35004.0       2.0    1350.0     630.0     360.0     300.0   

       PREP         N2  ...    N10300     A10300   N85530   A85530   N85300  \
0  455560.0  1356760.0  ...  373410.0   328469.0      0.0      0.0      0.0   
1  275920.0  1010990.0  ...  395880.0   965011.0      0.0      0.0      0.0   
2  155100.0   583910.0  ...  251490.0  1333418.0      0.0      0.0      0.0   
3   99950.0   423990.0  ...  165320.0  1414283.0      0.0      0.0      0.0   
4  126860.0   589490.0  ...  212000.0  3820152.0    420.0    168.0     60.0   
5   41410.0   160530.0  ...   55300.0  6027793.0  22090.0  39519.0  27550.0   
6     700.0     2160.0  ...     690.0      610.0      0.0      0.0      0.0   
7     610.0     2540.0  ...    1140.0     3019.0      0.0      0.0      0.0   

    A85300   N11901    A11901    N11902     A11902  
0      0.0  61920.0   48150.0  732670.0  1933120.0  
1      0.0  73720.0  107304.0  415410.0  1187403.0  
2      0.0  64200.0  139598.0  193030.0   536699.0  
3      0.0  45460.0  128823.0  116440.0   377177.0  
4     31.0  83330.0  421004.0  121570.0   483682.0  
5  95112.0  28590.0  791573.0   15960.0   250289.0  
6      0.0    120.0      94.0    1290.0     2792.0  
7      0.0    210.0     301.0    1130.0     2935.0  

[8 rows x 131 columns]

*  Modify `tax_data` to uppercase all the header names.
  * Your answer can only use `DataFrame` or `Series` methods or properties.
  * Do not hardcode the operation by uppercasing the column names yourself.
  * You can use `tax_data.columns`, which returns a `Series` of the column names.
  
* The resulting column names should look as follows:


```
STATEFIPS    STATE    ZIPCODE    AGI_STUB    N1    MARS1    MARS2    MARS4    PREP    N2    ...    
```

* This operation is useful for standardizing column names and avoid guessing whether the column header was in upper case, lower case, or a mix of both.

In [7]:
## WRITE YOUR CODE HERE

tax_data.columns = tax_data.columns.str.upper()
print(tax_data.columns)

Index(['STATEFIPS', 'STATE', 'ZIPCODE', 'AGI_STUB', 'N1', 'MARS1', 'MARS2',
       'MARS4', 'PREP', 'N2',
       ...
       'N10300', 'A10300', 'N85530', 'A85530', 'N85300', 'A85300', 'N11901',
       'A11901', 'N11902', 'A11902'],
      dtype='object', length=131)


* What is the total number of entries (also called observations) in `tax_data`?

  * Your answer can only use `DataFrame` or `Series` methods or properties.

In [9]:
## WRITE YOUR CODE HERE

tax_data.shape

(36637, 131)

- If `STATEFIPS` is header label of the first column (index 0) of the `tax_data` DataFrame, what is the label of the 32nd column?
  - You should use a single python expression.
  - Your answer can only use `DataFrame` or `Series` methods or properties.

In [12]:
## WRITE YOUR CODE HERE

tax_data.columns[31]

'N00900'

* If `STATEFIPS` is the first column label (index 0), what is the index of the column labeled `N10300`?
  * Your answer can only use `DataFrame` or `Series` methods or properties.

In [15]:
## WRITE YOUR CODE HERE

tax_data.columns.get_loc('N10300')

121

- Print the count of unique zip codes in each state using descending order.
  - Your answer can only use `DataFrame` or `Series` methods or properties.

- The result should look like the following. ( '...' represents remaining data that is not shown here.)

```python
        ZIPCODE
STATE
TX	     1619
NY	     1540
CA	     1485
PA	     1368
IL	     1231
...
```

* The above indicates that there are 1,619 unique zip codes in TX, 1,540 in NY, etc.

In [24]:
## WRITE YOUR CODE HERE

filtered_taxdata = tax_data.groupby('STATE')['ZIPCODE'].nunique().sort_values(ascending=False)

- Identify the position of `HI` in the list of zip code counts per state generated in the previous question.
  - Your answer can only use `DataFrame` or `Series` methods or properties.

In [26]:
## WRITE YOUR CODE HERE

print(filtered_taxdata.index.get_loc('HI'))

10


## 1.2. Identifying and removing entries with ambiguous zip codes

- Count and print the number of entries in `tax_data` where ZIPCODE is 0. Assign your results to a variable named  `nb_invalid_zip`.

In [29]:
## WRITE YOUR CODE HERE

filtered_taxdata = tax_data[tax_data['ZIPCODE'] == 0]
nb_invalid_zip = filtered_taxdata.shape[0]
print(nb_invalid_zip)

84


* Run the line below to make sure that `nb_invalid_zip` is an integer (`int`).
  * Note that `assert` will only print an error if `type(nb_invalid_zip)` is not of type `int`.
  
* If the code cell below returns an error, then change your answer above so that the value returned is effectively an integer.

In [30]:
assert(type(nb_invalid_zip) == int)

- Remove from `tax_data` all the lines where the zip code is `0` and save resulting `DataFrame` to a variable named `tax_data_valid_zip`.
  - Your answer can only use `DataFrame` or `Series` methods or properties.

In [35]:
## WRITE YOUR CODE HERE

tax_data_valid_zip = tax_data[tax_data['ZIPCODE'] != 0]

* Run the line below to confirm that the operation worked as expected.

  * The assertion below verifies that `nb_invalid_zip` + number of rows in `tax_data_valid_zip` is equal to the number of rows in the original DataFrame `tax_data`.
  
  * The assertion below will fail (and print an error message) if the results do not match. If that is the case, please review your code above.

In [36]:
assert((tax_data_valid_zip.shape[0] + nb_invalid_zip) == tax_data.shape[0])

## 1.3. Identifying and removing entries with missing values

* How many rows in the `tax_data_valid_zip` DataFrame contain at least one missing value (`NaN`) ?
  * Your answer can only use `DataFrame` methods and properties.
  
* Assign the count of such rows to a variable called `nb_missing_values`.

In [38]:
## WRITE YOUR CODE HERE

nb_missing_values = tax_data_valid_zip.isna().any(axis=1).sum()
print(nb_missing_values)

41


* Create a new `DataFrame` containing all the lines from `tax_data_valid_zip`, except lines containing missing values.

  * Call the new `DataFrame` `tax_data_valid_zip_cleaned`.

In [43]:
## WRITE YOUR CODE HERE

tax_data_valid_zip_cleaned = tax_data_valid_zip.dropna()
print(tax_data_valid_zip_cleaned)

       STATEFIPS STATE  ZIPCODE  AGI_STUB      N1  MARS1  MARS2  MARS4   PREP  \
6            1.0    AL  35004.0       1.0  1490.0  970.0  230.0  280.0  700.0   
7            1.0    AL  35004.0       2.0  1350.0  630.0  360.0  300.0  610.0   
8            1.0    AL  35004.0       3.0   970.0  310.0  490.0  140.0  450.0   
9            1.0    AL  35004.0       4.0   620.0  110.0  470.0   30.0  300.0   
10           1.0    AL  35004.0       5.0   620.0   40.0  560.0   20.0  320.0   
...          ...   ...      ...       ...     ...    ...    ...    ...    ...   
36631       17.0    IL  61540.0       2.0   360.0  180.0  120.0   70.0  210.0   
36632       17.0    IL  61540.0       3.0   210.0   70.0  130.0    0.0  130.0   
36633       17.0    IL  61540.0       4.0   140.0   40.0  110.0    0.0  100.0   
36634       17.0    IL  61540.0       5.0   160.0    0.0  140.0    0.0  120.0   
36635       17.0    IL  61540.0       6.0    30.0    0.0   20.0    0.0    0.0   

           N2  ...  N10300 

* Run the line below to confirm that the operation worked as expected. The assertion below tests that:  
`nb_missing_values` + number of rows in `tax_data_valid_zip_cleaned` is equal to the number of rows in `tax_data_valid_zip`.
  
* Note that `assert` will only print an error if the results do not match.

In [46]:
assert((tax_data_valid_zip_cleaned.shape[0] + nb_missing_values) == tax_data_valid_zip.shape[0])

## 1.4. Computing the percentile income per zip code

* The function `compute_percentile_zip` below computes the percentile income per zip code.
  * `N1` column: Number of tax returns
  * `A00100` column: Adjusted gross income (AGI)


* By default `percentile=0.5`,  i.e., the function computes the median.

* Read the code and make sure you understand what it does before moving on to the next question.


In [47]:
def compute_percentile_zip(df_zip, percentile=0.5):
    index_median = sum((df_zip["N1"] / sum(df_zip["N1"])).cumsum() <= percentile)
    val_below_or_at_median = (df_zip["A00100"] / df_zip["N1"]).iloc[index_median]
    return val_below_or_at_median

* Compute the 65th income percentile (percentile=0.65) for each zip code in `tax_data_valid_zip_cleaned` DataFrame.

  * Your answer can only use `DataFrame` or `Series` methods or properties.
  
  * Recall that you can use the `apply` function to transform the values of a column.
  
  
* Sort the values in descending order and assign them to a variable named `zip_rev_all`.
* The resulting `Series` should look like the following ( '...' represents remaining data that is not shown):

```Python
ZIPCODE
33109.0    3954.114286
33480.0    3413.3015383
...
```

In [49]:
## WRITE YOUR CODE HERE

zip_rev_all = tax_data_valid_zip_cleaned.groupby('ZIPCODE').apply(compute_percentile_zip, percentile=0.65).sort_values(ascending=False)
print(zip_rev_all)


ZIPCODE
33109.0    3954.114286
33480.0    3413.301538
94301.0    3109.443711
94027.0    3091.537013
90067.0    1986.449650
              ...     
33128.0      12.399005
95202.0      12.081356
32508.0      11.689423
32304.0      11.597207
61001.0      10.951744
Length: 6080, dtype: float64


<ipython-input-49-34f0cf5c2c4e>:3: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  zip_rev_all = tax_data_valid_zip_cleaned.groupby('ZIPCODE').apply(compute_percentile_zip, percentile=0.65).sort_values(ascending=False)


- What are the five zip codes with the most significant 65th percentile value for income?

In [50]:
## WRITE YOUR CODE HERE

zip_rev_all.head(5)

ZIPCODE
33109.0    3954.114286
33480.0    3413.301538
94301.0    3109.443711
94027.0    3091.537013
90067.0    1986.449650
dtype: float64

# 2. Working with the Medicaid Data

## 2.1. Loading and exploring the data

* Load the Medicaid data stored in the file `medicaid_data.csv` into a DataFrame called `medicaid_data`. The file is located in the `data` directory of the assignment folder.

* Note that this file is quite large and may take some time to load on a computer with modest RAM resources (e.g., 4GB or less).

In [51]:
## WRITE YOUR CODE HERE

medicaid_data = pd.read_csv('data/medicaid_data.csv')


- Modify `medicaid_data` to capitalize all the column names.

  - If your solution uses an assignment, the righthand side of the assignment (rvalue) can only use `DataFrame` or `Series` methods or properties.

In [52]:
## WRITE YOUR CODE HERE

medicaid_data.columns = medicaid_data.columns.str.upper()
print(medicaid_data.columns)

Index(['UTILIZATION TYPE', 'STATE', 'NDC', 'PRODUCT NAME', 'UNITS REIMBURSED',
       'NUMBER OF PRESCRIPTIONS', 'TOTAL AMOUNT REIMBURSED',
       'MEDICAID AMOUNT REIMBURSED', 'NON MEDICAID AMOUNT REIMBURSED',
       'LOCATION'],
      dtype='object')



- Familiarize yourself with the data.
  - The `NDC` column stands for National Drug Code, a universal product identifier for human drugs in the United States.
  
  - The remaining column names are self-explanatory.
  
  
- Explore the number of rows and columns in the data.

- Check that your column names are capitalized.

In [54]:
## WRITE YOUR CODE HERE

medicaid_data.shape

(1123003, 10)

* If you explore the  `LOCATION` column for all the entries for which `STATE` value is equal to "HI" you'll notice that all the values are identical. Confirm it.
  * Your answer can only use `DataFrame` or `Series` methods or properties.

In [60]:
## WRITE YOUR CODE HERE

medicaid_data[medicaid_data['STATE'] == 'HI']['LOCATION'].nunique()

1

* Are there any states that have more than one unique value for `LOCATION`? If there are, list those states.
  * This question can be answered by using aggregation and the Split-Apply-Combine paradigm.
  * Your answer can only use `DataFrame` or `Series` methods or properties.

In [64]:
## WRITE YOUR CODE HERE

location_counts = medicaid_data.groupby('STATE')['LOCATION'].nunique()

states_with_multiple_locations = location_counts[location_counts > 1]

# Display the results
print(states_with_multiple_locations)


Series([], Name: LOCATION, dtype: int64)




* To compare medication prescriptions across states in a fair and balanced way, we need the number of Medicaid beneficiaries in each state. That is, we need to compare per capita. The following example illustrates the importance of normalizing the values `UNITS REIMBURSED` for each medication in each state by the number of Medicaid enrollees in each state.
  
> The `medicaid_data` DataFrame shows that for the drug with NDC `61958180101` (the drug name is HARVONI and it's used to treat Hepatitis C) there were 11,886  units sold in KY, versus 40,142 in CA -- that's almost 4 times more units sold in CA compared to KY. However, there are 1,284,193 Medicaid enrollees in KY, versus 13,096,861 in California. If we normalize the number of units sold in KY versus CA, we find that there were close to 3 times more HARVONI prescription in KY  than in CA. This is ___perhaps___ justified by the fact the KY had one of the highest rates of reported cases of Hepatitis C in the US in 2015 (2.7% in KY versus 0.2% in CA).

    https://www.cdc.gov/hepatitis/statistics/2015surveillance/pdfs/2015hepsurveillancerpt.pdf

* The number of enrollees per state was obtained from [here](https://www.medicaid.gov/medicaid/managed-care/enrollment-report/index.html).


* A parsed/processed version (`medicaid_enrollment.tsv`) can be found in the `data` directory of the assignment folder.
* Use `pandas` to load the `medicaid_enrollmen.tsv` file into a DataFrame named `medicaid_enrollment`.

In [65]:
## WRITE YOUR CODE HERE

medicaid_enrollment = pd.read_csv('data/medicaid_enrollment.tsv', sep='\t')

* Modify `medicaid_enrollment` to capitalize all the column names.

  * Your answer can only use `DataFrame` or `Series` methods or properties.
  * Do not hardcode the operation by manually uppercasing the columns.

In [67]:
## WRITE YOUR CODE HERE

medicaid_enrollment.columns.str.upper()

Index(['STATE', 'TOTAL MEDICAID ENROLLEES'], dtype='object')

* Note that some states/territories have missing values. Remove the entries with missing values and save the resulting DataFrame as a new variable named `medicaid_enrollment_cleaned`.

* Pay attention to how 'n/a' is given here!
* After cleaning, do you still have the Guam entry? If so, reconsider what missing values means in this context.
    * Inspect the n/a values in `medicaid_enrollmen.tsv` file, and modify your code -- either here or the part reading in the file -- to make sure that the Guam entry doesn't appear after you remove the entries with missing values.
    * Do not hardcode it by manually deleting the Guam entry.

In [87]:
## WRITE YOUR CODE HERE

medicaid_enrollment_cleaned = medicaid_enrollment.dropna()
medicaid_enrollment_cleaned

STATE TOTAL MEDICAID ENROLLEES
0                    Alabama                1,050,989
1                     Alaska                  164,783
2             American Samoa                      nan
3                    Arizona                1,740,520
4                   Arkansas                  762,166
5                 California               13,096,861
6                   Colorado                1,264,600
7                Connecticut                  746,119
8                   Delaware                  227,909
9       District of Columbia                  271,428
10                   Florida                3,808,334
11                   Georgia                1,990,810
12                      Guam                      n/a
13                    Hawaii                  340,513
14                     Idaho                  283,355
15                  Illinois                3,269,999
16                   Indiana                1,295,358
17                      Iowa                  618,505
18                    Kansas                  403,844
19                  Kentucky                1,284,193
20                 Louisiana                1,402,212
21                     Maine                  288,324
22                  Maryland                1,271,445
23             Massachusetts                1,829,618
24                  Michigan                3,947,031
25                 Minnesota                1,052,521
26               Mississippi                  740,937
27                  Missouri                  944,257
28                   Montana                  139,950
29                  Nebraska                  239,463
30                    Nevada                  588,304
31             New Hampshire                  186,399
32                New Jersey                1,705,594
33                New Mexico                  826,155
34                  New York                6,281,038
35            North Carolina                1,965,805
36              North Dakota                   86,250
37  Northern Mariana Islands                      nan
38                      Ohio                3,060,446
39                  Oklahoma                  829,561
40                    Oregon                1,123,913
41              Pennsylvania                2,569,232
42               Puerto Rico                1,458,819
43              Rhode Island                  308,521
44            South Carolina                1,233,430
45              South Dakota                  124,497
46                 Tennessee                1,562,745
47                     Texas                4,273,982
48                      Utah                  293,867
49                   Vermont                  206,469
50            Virgin Islands                      nan
51                  Virginia                1,092,225
52                Washington                1,771,679
53             West Virginia                  545,748
54                 Wisconsin                1,209,714
55                   Wyoming                   66,532

## 2.2. Converting `TOTAL MEDICAID ENROLLEES` data type

* Given that data on `TOTAL MEDICAID ENROLLEES` column contains commas on file (ex. 3,269,999 instead of 3269999), `pandas` has erroneously set the data type for that column as `object`. We need to convert the column from `object` to `int` since we will be using it in an arithmetic expression during normalization.


* Convert the data type of `TOTAL MEDICAID ENROLLEES` column in `medicaid_enrollment_cleaned` to `int`.
* Inspect the `dtype` property of the column, and  make sure that the data type is `int`.

In [ ]:
## WRITE YOUR CODE HERE



## 2.3. Associating `medicaid_data` and `medicaid_enrollment_cleaned`

- We can use the shared State information across both tables to associate both tables (like SQL join).
- However, `medicaid_data` contains two-letter state abbreviations, while `medicaid_enrollment_cleaned` contains the complete state name.
  - We need to convert (or append) two-letter state abbreviations to `medicaid_enrollment_cleaned`.


- Pandas can read HTML and parse the code for tables. We will use that functionality to read in the state abbreviations from a web page.
  - A brief description of what the code does is included in the comments.
  
  
- Note that the code below may throw an error if `lxml` is not installed on your machine. You should install any Python packages that Python complains about before proceeding.

In [ ]:
# Reads all the tables at the given url.
# Header=0 instructs pandas to use line 0 as the header (columns)
import requests
from io import StringIO

url = requests.get('https://www.50states.com/abbreviations.htm')
tables = pd.read_html(StringIO(url.text), header=0)


# We access the desired table by giving its index.
# Since the URL contains three tables, we can access state abbreviations table using index 0.
Codes_abbreviations = tables[0]
Codes_abbreviations.head()

* Change the the `DataFrame`'s headers from ['US STATE',	'POSTAL ABBREVIATION', 	'STANDARD ABBREVIATION'] to ['US STATE', 'ABBREVIATION', 'STD ABBREVIATION']

  * You can hardcode this operation.

In [ ]:
## WRITE YOUR CODE HERE



* Combine the tables `medicaid_enrollment_cleaned` and `Codes_abbreviations` such that the resulting DataFrame contains all the columns in `medicaid_enrollment_cleaned` and only `ABBREVIATION` from `Codes_abbreviations`.
  * Use the **default option (`inner`)** to merge.
  * The resulting DataFrame should contain 50 entries.

* Save the results to a variable named `medicaid_enrollment_cleaned_with_zip`.

* `medicaid_enrollment_cleaned_with_zip` should look like the following ( '...' represents remaining data that is not shown):


```
        STATE   TOTAL MEDICAID ENROLLEES   ABBREVIATION
0     Alabama                    1050989             AL
1      Alaska                     164783             AK
2     Arizona                    1740520             AZ
3    Arkansas                     762166             AR
4  California                   13096861             CA
...
```

In [ ]:
## WRITE YOUR CODE HERE



* We have no further use for the column STATE in  `medicaid_enrollment_cleaned_with_zip`.
  * Remove the column. Make sure your data in `medicaid_enrollment_cleaned_with_zip` looks like the following  ( '...' represents remaining data that is not shown):

```
   TOTAL MEDICAID ENROLLEES   ABBREVIATION
0                   1050989             AL
1                    164783             AK
2                   1740520             AZ
3                    762166             AR
4                  13096861             CA
...
```

In [ ]:
## WRITE YOUR CODE HERE



- Use the DataFrame `medicaid_enrollment_cleaned_with_zip` to assign the appropriate number of Medicaid enrollees to each entry in the `medicaid_data`.

  - I.e., instead of the 10 original columns, `medicaid_data` will now have an 11th column representing the `TOTAL MEDICAID ENROLLEES` according to the `STATE` value in the entry.
  - Use the **default option (`inner`)** to merge.

   
- Save the resulting DataFrame into a new variable called `medicaid_data_w_enrollments`.
- The resulting DataFrame should look like the following ( '...' represents remaining data that is not shown):

```
                                                           NON                  TOTAL
  UTILIZATION                         PRODUCT  ...    MEDICAID               MEDICAID
         TYPE  STATE          NDC        NAME           AMOUNT    LOCATION  ENROLLEES
                                                    REIMBURSED
0        MCOU     PA  55150023930  Dexamethas  ...        0.00    (40.5773,  2569232.0
                                                                   -77.264)
1        FFSU     NY     23917710  ALPHAGAN P  ...        0.00    (42.1497,  6281038.0
                                                                  -74.9384)
2        MCOU     OR  13925050501  Dapsone 10  ...        0.00    (44.5672,  1123913.0
                                                                 -122.1269)      
...
```

* The order of the columns in the DataFrame is not important. This answer uses the same approach as the one used to `merge` the tables above.

In [ ]:
## WRITE YOUR CODE HERE



- Remove any rows where `STATE` or `PRODUCT NAME` are missing from  `medicaid_data_w_enrollments`.

In [ ]:
## WRITE YOUR CODE HERE



- Use `["STATE", "NDC"]` as hierarchical index for `medicaid_data_w_enrollments`. Recall that a hierarchical index is simply an index with multiple levels of indexing (multiple columns).
  * Hint: The function to set an index on a DataFrame can take a single column name or a list of column names. The list here is `["STATE", "NDC"]`.
  
  
- Call the new data `medicaid_data_w_enrollments_hierarch`.


- Inspect your data to make sure the new index has now two levels (STATE and NDC).

In [ ]:
## WRITE YOUR CODE HERE




* Write a single Pandas expression to show all the rows containing `NDC` `61958180101` in the Pennsylvania ("PA").

 * Use a single indexing call (bracket notation) using `loc`.

 * Hint: Since your index is hierarchical, `loc` is expecting two values, the first for `STATE` and the second for `NDC`.

In [ ]:
## WRITE YOUR CODE HERE



## 2.4. Computing the normalized `UNITS REIMBURSED` per enrollee

* Compute the `UNITS REIMBURSED` for each "NDC" in each state normalized by the number of enrollees.


> - For instance in `PA`, the total `UNITS REIMBURSED` for the HARVONI `NDC` (61958180101) is 10,612.

```python
      total_amount_reimbursed = medicaid_data_w_enrollments_hierarch.loc[("PA", 61958180101), "UNITS REIMBURSED"].sum()
```

> - And the numebr of Medicaid Enrollees in "PA" is 2,569,232.

```python
      total_enrollees_PA = medicaid_data_w_enrollments_hierarch.loc["PA", "TOTAL MEDICAID ENROLLEES"].unique()[0]
```

> - Therefore, the `UNITS REIMBURSED` per enrollee for "HARVONI" is  0.00413041718303.

```python
      print(total_amount_reimbursed/total_enrollees_PA)
```

- Rather than work directly with the ratios, we are going to compute and work with their logarithm (`np.log2`) instead.

  - The reason we use logs here is to avoid working small numbers. More on log-transformation in future lectures.
  

- Save the result in a `Series` called `medicaid_reimbursement_per_enrollee`.


- Your final result should be a `Series` that look like the following ( '...' represents data that is not shown here):

  - Note that negative values are the result of the `log` transformation.
  
```
STATE  NDC    
AK     2143380    -9.609109
       2143480   -10.008280
       2322730    -6.109830
       2322830    -4.444321
       2322930    -3.855995
...

AL     2143380    -9.940595
       2143480    -9.805485
       2322730    -5.336260
       2322830    -4.243688
       2322930    -3.645575

...

MA     2143380    -7.921997
       2143480    -7.803463
       2144511   -13.472194
       2197590    -7.741402
       2322730    -5.414724
       2322830    -4.592154
       2322930    -4.205626
...
```

In [ ]:
## WRITE YOUR CODE HERE



- To facilitate working with the final data, we are going to unstack `medicaid_reimbursement_per_enrollee` into a variable called  `medicaid_norm_ndc`.


- Using `medicaid_reimbursement_per_enrollee`, generate a DataFrame where:
  - index should be the two-letter state symbol
  - column names are the `NDC` codes


- The DataFrame should look like the following ('...' represents data that is not shown):
  - Hint: Simply unstack the data.
  
  
```
  NDC    2143380     2143480     2144511 2144527    2197590    2322730 ... 99207026012 99207046330
STATE
   AK  -9.609109  -10.008280         NaN     NaN        NaN  -6.109830 ...         NaN         NaN
   AL  -9.940595   -9.805485         NaN     NaN        NaN  -5.336260 ...         NaN         NaN
   AR -12.985157  -12.657103         NaN     NaN        NaN  -4.720964 ...         NaN         NaN
   AZ -11.533870  -10.821194         NaN     NaN        NaN  -5.312806 ...         NaN         NaN
   CA -10.926863  -10.698372  -16.127018     NaN  -9.464364  -7.429936 ...         NaN  -15.150865
...
```

In [ ]:
## WRITE YOUR CODE HERE



## 2.5. Exploring the data (very briefly)
- What is the NDC for the drug with the highest log-normalized ratio in Hawaii?

In [ ]:
## WRITE YOUR CODE HERE



* Investigate the NDC of the product with the highest log-normalized ratio in Hawaii.
  * What is the name of the product? What is it used for?

In [ ]:
## WRITE YOUR CODE HERE



> **WRITE YOUR ANSWER HERE:**
>
>


* Compare the log-normalized `UNITS REIMBURSED` per enrollee of that product between HI and other states, (take for instance MA, FL, OR and WA).

In [ ]:
## WRITE YOUR CODE HERE



* Can you think for reasons why this product has the highest log-normalized ratio in Hawaii? (Not graded)

> **WRITE YOUR ANSWER HERE**
>
>